lib/Dialect/FIRRTL/Transforms/DropConst.cpp

```diff
@@ -0,0 +1,184 @@
+
+    if (isWithinNonconstCondition)
+      nonConstConditionedBlock = &when.getThenBlock();
+    for (auto &op : llvm::make_early_inc_range(when.getThenBlock()))
+      dispatchVisitor(&op);
+
+    if (when.hasElseRegion()) {
+      if (isWithinNonconstCondition)
+        nonConstConditionedBlock = &when.getElseBlock();
+      for (auto &op : llvm::make_early_inc_range(when.getElseBlock()))
+        dispatchVisitor(&op);
+    }
+
+    nonConstConditionedBlock = previousNonConstConditionedBlock;
+  }
+
+  void visitStmt(ConnectOp connect) { handleConnect(connect); }
+
+  void visitStmt(StrictConnectOp connect) { handleConnect(connect); }
+
+  void visitUnhandledOp(Operation *op) {
+    // Register any 'const' results to drop 'const'
+    for (auto [index, result] : llvm::enumerate(op->getResults())) {
+      if (auto convertedType = convertType(result.getType()))
+        constValuesToConvert.push_back({result, convertedType});
+    }
+  }
+
+private:
+  /// Returns null type if no conversion is needed
+  Type convertType(Type type) {
+    if (auto base = type.dyn_cast<FIRRTLBaseType>()) {
+      return convertType(base);
+    }
+
+    if (auto refType = type.dyn_cast<RefType>()) {
+      if (auto converted = convertType(refType.getType()))
+        return RefType::get(converted.cast<FIRRTLBaseType>(),
+                            refType.getForceable());
+    }
+
+    return {};
+  }
+
+  /// Returns null type if no conversion is needed
+  FIRRTLBaseType convertType(FIRRTLBaseType type) {
+    auto nonConstType = type.getAllConstDroppedType();
+    return nonConstType != type ? nonConstType : FIRRTLBaseType{};
+  }
+
+  void handleConnect(FConnectLike connect) {
+    auto dest = connect.getDest();
+    auto destType = dest.getType();
+    if (nonConstConditionedBlock && containsConst(destType)) {
+      // 'const' connects are allowed if `dest` is local to the non-'const' when
+      // block
+      auto *destBlock = dest.getParentBlock();
+      auto *block = connect->getBlock();
+      while (block) {
+        // The connect is local to the `dest` declaration, both local to the
+        // non-'const' block, so the connect is valid
+        if (block == destBlock)
+          return;
+        // The connect is within the non-'const' condition, non-local to the
+        // `dest` declaration, so the connect is invalid
+        if (block == nonConstConditionedBlock)
+          break;
+
+        if (auto *parentOp = block->getParentOp())
+          block = parentOp->getBlock();
+        else
+          break;
+      }
+
+      result = failure();
+      if (isConst(destType))
+        connect.emitOpError() << "assignment to 'const' type " << destType
+                              << " is dependent on a non-'const' condition";
+      else
+        connect->emitOpError()
+            << "assignment to nested 'const' member of type " << destType
+            << " is dependent on a non-'const' condition";
```

Comment:
As a general rule, we break after the first error - we never print all possible errors.  What we really need is a framework to limit the number of errors printed to some maximum.

Reply:
Updated to only allow for one error by returning `LogicalResult` from the visit functions